src/optimizer/rule/PushFilterDownProjectRule.cpp

```diff
@@ -0,0 +1,72 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "optimizer/rule/PushFilterDownProjectRule.h"
+
+#include "optimizer/OptContext.h"
+#include "optimizer/OptGroup.h"
+#include "planner/plan/PlanNode.h"
+#include "planner/plan/Query.h"
+#include "util/ExpressionUtils.h"
+
+using nebula::graph::PlanNode;
+using nebula::graph::QueryContext;
+
+namespace nebula {
+namespace opt {
+
+std::unique_ptr<OptRule> PushFilterDownProjectRule::kInstance =
+    std::unique_ptr<PushFilterDownProjectRule>(new PushFilterDownProjectRule());
+
+PushFilterDownProjectRule::PushFilterDownProjectRule() {
+    RuleSet::QueryRules().addRule(this);
+}
+
+const Pattern& PushFilterDownProjectRule::pattern() const {
+    static Pattern pattern = Pattern::create(graph::PlanNode::Kind::kFilter,
+                                             {Pattern::create(graph::PlanNode::Kind::kProject)});
+    return pattern;
+}
+
+StatusOr<OptRule::TransformResult> PushFilterDownProjectRule::transform(
+    OptContext* octx,
+    const MatchedResult& matched) const {
+    auto filterGroupNode = matched.node;
+    auto oldFilterNode = filterGroupNode->node();
+    auto projGroupNode = matched.dependencies.front().node;
+    auto oldProjNode = projGroupNode->node();
+
+    auto newFilterNode = static_cast<graph::Filter*>(oldFilterNode->clone());
+    auto newProjNode = static_cast<graph::Project*>(oldProjNode->clone());
+
+    // Exchange planNode
+    newProjNode->setOutputVar(oldFilterNode->outputVar());
```

Comment:
Non-VariablePropertyExpressions may need to be rewritten. 


Before:
`Project : a,abs(b) AS b,c+1 AS c => Filter: a>1 and b-1>3 and c>3`
After:
`Filter : a>1 and abs(b)-1>3 and c+1>3  => Project: a, abs(b) AS b, c+1 AS c`